pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,284 @@
+	rev          RevisionID
+	updateCh     chan<- *RevisionDestsUpdate
+	dests        []string
+	healthStates map[string]bool
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev RevisionID, updateCh chan<- *RevisionDestsUpdate,
+	destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	return &revisionWatcher{
+		rev:          rev,
+		updateCh:     updateCh,
+		healthStates: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			// Prefer IP as we can avoid a DNS lookup this way
+			ret = append(ret, net.JoinHostPort(addr.IP, strconv.Itoa(int(portVal))))
+		}
+	}
+
+	return ret
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := []string{}
+	for dest, healthy := range dests {
+		if healthy {
+			ret = append(ret, dest)
+		}
+	}
+	return ret
+}
+
+type destHealth struct {
+	dest   string
+	health bool
+}
+
+func (rw *revisionWatcher) checkDests() {
+	healthStatesCh := make(chan *destHealth, len(rw.dests))
+	defer close(healthStatesCh)
+
+	// Context used for our requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthStates[dest]; ok && curHealthy {
+			healthStatesCh <- &destHealth{dest, true}
+			continue
+		}
+
+		go func(dest string) {
+			httpDest := url.URL{
+				Scheme: "http",
+				Host:   dest,
+			}
+			ok, err := prober.Do(ctx, rw.transport, httpDest.String(),
+				prober.WithHeader(network.ProbeHeaderName, queue.Name),
+				prober.ExpectsBody(queue.Name))
+
+			healthStatesCh <- &destHealth{dest, ok}
+
+			if err != nil {
+				rw.logger.Errorw("Failed probing "+dest, zap.Error(err))
+			} else if !ok {
+				rw.logger.Info("Probing of dest " + dest)
+			}
+		}(dest)
+	}
+
+	healthStates := make(map[string]bool)
+	for i := 0; i < len(rw.dests); i++ {
+		dh := <-healthStatesCh
+		healthStates[dh.dest] = dh.health
+	}
+
+	if !reflect.DeepEqual(rw.healthStates, healthStates) {
+		rw.healthStates = healthStates
+		destsArr := filterHealthyDests(healthStates)
+		rw.updateCh <- &RevisionDestsUpdate{
+			Rev:   rw.rev,
+			Dests: destsArr,
+		}
+	}
+}
+
+func (rw *revisionWatcher) runWithTickCh(tickCh <-chan time.Time) {
+	for {
+		select {
+		case dests, ok := <-rw.destsChan:
+			if !ok {
+				// shutdown
+				return
+			}
+			rw.dests = dests
+			rw.checkDests()
```

Comment:
you can just move `rw.checkDests()` outside the select, since in both cases this is executed (select basically becomes the timer/interrupt).